### PR TITLE
[query] Don't create java objects in TableParallelize

### PR DESCRIPTION
**Pull request by tpoterba**
Don't love allocating input/outputstreams per row, but still better than before.

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/ir/TableIR.scala [hunk 1]
```diff
@@ -326,26 +326,41 @@ case class TableParallelize(rowsAndGlobal: IR, nPartitions: Option[Int] = None)
...
 
-    val encRows = Array.tabulate(nRows) { i =>
-      if (rowsT.isElementMissing(rowsAddr, i))
-        fatal(s"cannot parallelize null values: found null value at index $i")
-      val baos = new ByteArrayOutputStream()
-      makeEnc(baos).writeRegionValue(ctx.r, rowsT.loadElement(rowsAddr, i))
-      baos.toByteArray
+    val nSplits = nPartitions.getOrElse(16)
+    val rowsPerPartition = nRows / nSplits
+    val leftOver = nRows - (nSplits * rowsPerPartition) // if partition index < leftOver, has rowsPerPartition + 1
```
Comment (cseed) [inline hail/src/main/scala/is/hail/expr/ir/TableIR.scala:331]: nSplits shouldn't be bigger than nRows.

You can use `utils.partition` to compute partition splits.  It's basically what you do here, except it has been debugged and has some nice checks.  The idiom we use in most places is:

```
        val partCounts = partition(n, nPartitions)
        val partStarts = partCounts.scanLeft(0)(_ + _)
```

File: hail/src/main/scala/is/hail/expr/ir/TableIR.scala [hunk 2]
```diff
@@ -326,26 +326,41 @@ case class TableParallelize(rowsAndGlobal: IR, nPartitions: Option[Int] = None)
...
       .cmapPartitions { (ctx, it) =>
         val rv = RegionValue(ctx.region)
-        it.map { arr =>
-          val bais = new ByteArrayInputStream(arr)
-          rv.setOffset(makeDec(bais).readRegionValue(ctx.region))
-          rv
+        it.flatMap { case (nRows, arr) =>
+          val bais = new ByteArrayDecoder(makeDec)
+          bais.set(arr)
+          Iterator.range(0, nRows)
```
Comment (cseed) [inline hail/src/main/scala/is/hail/expr/ir/TableIR.scala:359]: Doesn't this need to be the number of elements in this partition?

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/TableIR.scala:359]: it is...bad variable name (overloaded)

File: hail/src/main/scala/is/hail/expr/ir/TableIR.scala (hunk 1 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/TableIR.scala:331]: ah, I see. this is still correct, but can result in 0-element partitions.

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/TableIR.scala:331]: will switch.

File: hail/src/main/scala/is/hail/expr/ir/TableIR.scala [hunk 3]
```diff
@@ -311,25 +311,58 @@ case class TableParallelize(rowsAndGlobal: IR, nPartitions: Option[Int] = None)
...
+
+    val rowsT = ptype.types(0).asInstanceOf[PArray]
+    val rowT = rowsT.elementType.asInstanceOf[PStruct]
+    val spec = TypedCodecSpec(rowT, BufferSpec.wireSpec)
+
+    val makeEnc = spec.buildEncoder(rowT)
+    val rowsAddr = ptype.loadField(res, 0)
+    val nRows = rowsT.loadLength(rowsAddr)
+
+    val parts = partition(nRows, nPartitions.getOrElse(16))
```
Comment (cseed) [inline hail/src/main/scala/is/hail/expr/ir/TableIR.scala:329]: I still think you should min with nRows.

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/TableIR.scala]: oops, I had a brain fart and assumed `partition` was doing that. Fixed.

**Pull request merged**